crates/rpc/rpc-builder/src/rate_limiter.rs

```diff
@@ -0,0 +1,109 @@
+use jsonrpsee::{server::middleware::rpc::RpcServiceT, types::Request, MethodResponse};
+use reth_tasks::pool::BlockingTaskGuard;
+use tokio::sync::OwnedSemaphorePermit;
+use std::{
+    future::Future,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+use tower::Layer;
+
+/// Rate limiter for the RPC server.
+///
+/// Rate limits expensive calls such as debug_ and trace_.
+#[derive(Debug, Clone)]
+pub(crate) struct RpcRequestRateLimiter {
+    inner: Arc<RpcRequestRateLimiterInner>,
+}
+
+impl RpcRequestRateLimiter {
+    pub(crate) fn new(handle: tokio::runtime::Handle, rate_limit: usize) -> Self {
+        Self {
+            inner: Arc::new(RpcRequestRateLimiterInner {
+                handle,
+                call_guard: BlockingTaskGuard::new(rate_limit),
+            }),
+        }
+    }
+}
+
+impl<S> Layer<S> for RpcRequestRateLimiter {
+    type Service = RpcRequestRateLimitingService<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        RpcRequestRateLimitingService::new(inner, self.clone())
+    }
+}
+
+/// Rate Limiter for the RPC server
+#[derive(Debug)]
+struct RpcRequestRateLimiterInner {
+    call_guard: BlockingTaskGuard,
+    handle: tokio::runtime::Handle,
+}
+
+/// A [`RpcServiceT`] middleware that rate limits RPC calls for the server.
+///
+#[derive(Clone, Debug)]
+pub struct RpcRequestRateLimitingService<S> {
+    /// The rate limiter for RPC requests
+    rate_limiter: RpcRequestRateLimiter,
+    /// The inner service being wrapped
+    inner: S,
+}
+
+impl<S> RpcRequestRateLimitingService<S> {
+    pub(crate) fn new(service: S, rate_limiter: RpcRequestRateLimiter) -> Self {
+        Self { inner: service, rate_limiter }
+    }
+}
+
+impl<'a, S> RpcServiceT<'a> for RpcRequestRateLimitingService<S>
+where
+    S: RpcServiceT<'a> + Send + Sync + Clone + 'static,
+{
+    type Future = RateLimitingRequestFuture<S::Future>;
+
+    fn call(&self, req: Request<'a>) -> Self::Future {
+        let method_name = req.method_name();
+        if method_name.starts_with("trace_") || method_name.starts_with("debug_") {
```

Comment:
I know in the issue it mentioned hashmap or filter, but was wondering if this is okay as well?

Reply:
this would be fine for now but to make this general purpose ratelimit filter we can also introduce something like a filter trait predicate that we can then, like `Arc< dyn RequestMatcher>`
`trait RequestMatcher  {fn is_match(&self , &Request) -> bool}`
or similar